src/tools/ceph-dencoder/rbd_types.h

```diff
@@ -3,13 +3,26 @@
 TYPE(librbd::journal::EventEntry)
 TYPE(librbd::journal::ClientData)
 TYPE(librbd::journal::TagData)
+using namespace librbd::journal;
+TYPE(EventEntry)
+TYPE(ClientData)
+TYPE(TagData)
+
 #include "librbd/mirroring_watcher/Types.h"
 TYPE(librbd::mirroring_watcher::NotifyMessage)
 #include "librbd/trash_watcher/Types.h"
 TYPE(librbd::trash_watcher::NotifyMessage)
 #include "librbd/WatchNotifyTypes.h"
 TYPE_NOCOPY(librbd::watch_notify::NotifyMessage)
+TYPE(librbd::watch_notify::AsyncRequestId)
 TYPE(librbd::watch_notify::ResponseMessage)
+using namespace librbd::watch_notify;
+TYPE_NOCOPY(NotifyMessage)
+TYPE(ResponseMessage)
+TYPE(AsyncRequestId)
```

Comment:
Same here and below for `using namespace cls::rbd;`.  I don't understand why these entries are being duplicated.

Reply:
> the corpuse repository contain types without namespaces.

Still not following.  For example, RBD has several different `NotifyMessage` types in different namespaces.  How does that work?

> that causing ceph-dencoder to output the type is not recognize (wihtout full namespace)

Do you have an example of such output?

Reply:
got it, so we will need from now to change the WRITE_CLASS_ENCODER so we will dump the type with the namespace, currently we have WRITE_CLASS_ENCODER(NotifyMessage) so it will be WRITE_CLASS_ENCODER(librbd::watch_notify::NotifyMessage).

we will lose the archived version checks, but we will make it right for the next versions.

Reply:
> got it, so we will need from now to change the WRITE_CLASS_ENCODER so we will dump the type with the namespace, currently we have WRITE_CLASS_ENCODER(NotifyMessage) so it will be WRITE_CLASS_ENCODER(librbd::watch_notify::NotifyMessage).

My first instinct would be to change `WRITE_CLASS_ENCODER` macro to get the fully-qualified name of the type internally and use that for naming the dump file.  But, if

> we will lose the archived version checks, but we will make it right for the next versions.

is the case, that is probably too disruptive.

Hrm...  I think `NotifyMessage` is unique in that we have five of them in different namespaces so as a (rather poor) workaround we might consider just giving each of them a unique name.

But fully-qualified names have been there in ceph-dencoder `types.h` and later `rbd_types.h` for years: see e.g. https://github.com/ceph/ceph/commit/eb8d37a105d0890e50f0844cd407ad114224255f.  Are you saying that it never worked?

Can you describe the breakage related to the object corpus in detail?  Since three out of five `NotifyMessage` types happen to be registered with ceph-dencoder but their fully-qualified names get stripped, what actually made it to the corpus?

Reply:
@idryomov thats not what i ment, it worked, but the corpus couldn't test it, when we keep the object for each version we are using DENCODE_DUMP, which is keeping the type name using WRITE_CLASS_ENCODER, in that case we are not able to tell (during corpus test) if the NotifyMessage came from watch_notify\trash_watcher or mirroring_watcher
now with the change of WRITE_CLASS_ENCODER with full namespace we will be able to tell and use the correct ceph-dencoder type

Reply:
So is it that only `NotifyMessage` types are affected or all types in `rbd_types.h` are affected (because fully-qualified names have always been used there)?

Reply:
Also: 
using namespace librbd::journal;
TYPE(EventEntry)
TYPE(ClientData)
TYPE(TagData)
I fixed the WRITE_CLASS_ENCODER for them as well

Reply:
Then I think I'm still missing something.  Why is e.g. `EventEntry` affected but e.g. `ResponseMessage` isn't?  They are both unqualified for `WRITE_CLASS_ENCODER`:
```
src/librbd/journal/Types.h:WRITE_CLASS_ENCODER(EventEntry);
src/librbd/WatchNotifyTypes.h:WRITE_CLASS_ENCODER(ResponseMessage);
```
and fully-qualified for `TYPE`:
```
TYPE(librbd::journal::EventEntry)
TYPE(librbd::watch_notify::ResponseMessage)
```

Reply:
why do you think ResponseMessage is not affected? i also added namespace for it .. but removed it after you pointed this out

Reply:
> why do you think ResponseMessage is not affected?

Because when I asked 

> So is it that only `NotifyMessage` types are affected or all types in `rbd_types.h` are affected (because fully-qualified names have always been used there)?

you responded with

> Also:
> using namespace librbd::journal;
> TYPE(EventEntry)
> TYPE(ClientData)
> TYPE(TagData)
> I fixed the WRITE_CLASS_ENCODER for them as well

and not with "no, `ResponseMessage` is also affected" or with "no, all types in `rbd_types.h` are affected".

To get us on the same page, let me restate my questions from https://github.com/ceph/ceph/pull/52481#discussion_r1275299259:
1. Is it true that none of the types in `rbd_types.h` have ever been stored/tested correctly in the corpus?
2. If (1) is true, is it because `rbd_types.h` (and RBD-related types in `types.h` before it got split) have always had fully-qualified names?
3. If (1) is false, which types in `rbd_types.h` are stored/tested correctly in the corpus and which aren't?

Reply:
you are right, my mistake, i didn't mention that correctly.

1. false. Most of the types were tested correctly.
3. since the corpus script first check if the type is supported by ceph-dencoder, its ignore all the types that return unsupported. that means that ResponseMessage \ NotifyMessage \  AsyncRequestId \ EventEntry \ ClientData \ TagData - returned unsupported and were never tested.

with the recent change for WRITE_CLASS_ENCODER for those types, we will dump the correct name of the type with namespace included and we will start test it.

Reply:
@rzarzynski has indicated that there is still interest in these ceph-dencoder PRs, so I'm taking another look.

> > Is it true that none of the types in rbd_types.h have ever been stored/tested correctly in the corpus?

> false. Most of the types were tested correctly.

Are you sure about the "most of the types" part?  I put together a list of types that are related to RBD:

```
WRITE_CLASS macro in the code                                                               TYPE macro in rbd_types.h or as noted                       STATUS      NOTES

src/cls/journal/cls_journal_types.h:WRITE_CLASS_ENCODER(ObjectPosition);                    TYPE(cls::journal::ObjectPosition)                                      in rgw_types.h -- why?
src/cls/journal/cls_journal_types.h:WRITE_CLASS_ENCODER(ObjectSetPosition);                 TYPE(cls::journal::ObjectSetPosition)                                   in rgw_types.h -- why?
src/cls/journal/cls_journal_types.h:WRITE_CLASS_ENCODER(Client);                            TYPE(cls::journal::Client)                                              in rgw_types.h -- why?
src/cls/journal/cls_journal_types.h:WRITE_CLASS_ENCODER(Tag);                               TYPE(cls::journal::Tag)                                                 in rgw_types.h -- why?
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_lock_op)                           TYPE(cls_lock_lock_op)                                      MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_unlock_op)                         TYPE(cls_lock_unlock_op)                                    MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_break_op)                          TYPE(cls_lock_break_op)                                     MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_get_info_op)                       TYPE(cls_lock_get_info_op)                                  MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER_FEATURES(cls_lock_get_info_reply)           TYPE_FEATUREFUL(cls_lock_get_info_reply)                    MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_list_locks_reply)                  TYPE(cls_lock_list_locks_reply)                             MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_assert_op)                         TYPE(cls_lock_assert_op)                                    MATCH       in common_types.h
src/cls/lock/cls_lock_ops.h:WRITE_CLASS_ENCODER(cls_lock_set_cookie_op)                     TYPE(cls_lock_set_cookie_op)                                MATCH       in common_types.h
src/cls/lock/cls_lock_types.h:      WRITE_CLASS_ENCODER(locker_id_t)                        TYPE(rados::cls::lock::locker_id_t)                                     in common_types.h
src/cls/lock/cls_lock_types.h:      WRITE_CLASS_ENCODER_FEATURES(locker_info_t)             TYPE_FEATUREFUL(rados::cls::lock::locker_info_t)                        in common_types.h
src/cls/lock/cls_lock_types.h:      WRITE_CLASS_ENCODER_FEATURES(lock_info_t);              TYPE_FEATUREFUL(rados::cls::lock::lock_info_t)                          in common_types.h
src/cls/rbd/cls_rbd.h:WRITE_CLASS_ENCODER_FEATURES(cls_rbd_parent)                          TYPE_FEATUREFUL(cls_rbd_parent)                             MATCH
src/cls/rbd/cls_rbd.h:WRITE_CLASS_ENCODER_FEATURES(cls_rbd_snap)                            TYPE_FEATUREFUL(cls_rbd_snap)                               MATCH
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MirrorPeer);                                TYPE(cls::rbd::MirrorPeer)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MirrorImage);                               TYPE(cls::rbd::MirrorImage)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MirrorImageSiteStatus);                     TYPE(cls::rbd::MirrorImageSiteStatus)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER_FEATURES(MirrorImageSiteStatusOnDisk)       TYPE_FEATUREFUL(cls::rbd::MirrorImageSiteStatusOnDisk)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MirrorImageStatus);                         TYPE(cls::rbd::MirrorImageStatus)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(ParentImageSpec);                           TYPE(cls::rbd::ParentImageSpec)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(ChildImageSpec);                            TYPE(cls::rbd::ChildImageSpec)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(GroupImageSpec);                            TYPE(cls::rbd::GroupImageSpec)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(GroupImageStatus);                          TYPE(cls::rbd::GroupImageStatus)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(GroupSpec);                                 TYPE(cls::rbd::GroupSpec)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(SnapshotNamespace);                         TYPE(cls::rbd::SnapshotNamespace)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(SnapshotInfo);                              TYPE(cls::rbd::SnapshotInfo)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(ImageSnapshotSpec);                         TYPE(cls::rbd::ImageSnapshotSpec)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(GroupSnapshot);                             TYPE(cls::rbd::GroupSnapshot)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(TrashImageSpec);
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MirrorImageMap);                            TYPE(cls::rbd::MirrorImageMap)
src/cls/rbd/cls_rbd_types.h:WRITE_CLASS_ENCODER(MigrationSpec);                             TYPE(cls::rbd::MigrationSpec)
src/journal/Entry.h:WRITE_CLASS_ENCODER(journal::Entry)                                     TYPE(journal::Entry)                                        MATCH       in common_types.h
src/librbd/WatchNotifyTypes.h:WRITE_CLASS_ENCODER(ClientId);                                                                                                        duplicate WRITE_CLASS_ENCODER for librbd::watcher::ClientId
src/librbd/WatchNotifyTypes.h:WRITE_CLASS_ENCODER(AsyncRequestId);
src/librbd/WatchNotifyTypes.h:WRITE_CLASS_ENCODER(NotifyMessage);                           TYPE_NOCOPY(librbd::watch_notify::NotifyMessage)
src/librbd/WatchNotifyTypes.h:WRITE_CLASS_ENCODER(ResponseMessage);                         TYPE(librbd::watch_notify::ResponseMessage)
src/librbd/cache/pwl/Types.h:WRITE_CLASS_DENC(librbd::cache::pwl::WriteLogCacheEntry)       TYPE(librbd::cache::pwl::WriteLogCacheEntry)                MATCH
src/librbd/cache/pwl/Types.h:WRITE_CLASS_DENC(librbd::cache::pwl::WriteLogPoolRoot)         TYPE(librbd::cache::pwl::WriteLogPoolRoot)                  MATCH
src/librbd/cache/pwl/ssd/Types.h:WRITE_CLASS_DENC(librbd::cache::pwl::ssd::SuperBlock)      TYPE(librbd::cache::pwl::ssd::SuperBlock)                   MATCH
src/librbd/journal/Types.h:WRITE_CLASS_ENCODER(EventEntry);                                 TYPE(librbd::journal::EventEntry)
src/librbd/journal/Types.h:WRITE_CLASS_ENCODER(ClientData);                                 TYPE(librbd::journal::ClientData)
src/librbd/journal/Types.h:WRITE_CLASS_ENCODER(TagData);                                    TYPE(librbd::journal::TagData)
src/librbd/mirror/snapshot/Types.h:WRITE_CLASS_ENCODER(ImageStateHeader);
src/librbd/mirror/snapshot/Types.h:WRITE_CLASS_ENCODER(SnapState);
src/librbd/mirror/snapshot/Types.h:WRITE_CLASS_ENCODER(ImageState);
src/librbd/mirroring_watcher/Types.h:WRITE_CLASS_ENCODER(NotifyMessage);                    TYPE(librbd::mirroring_watcher::NotifyMessage)
src/librbd/trash_watcher/Types.h:WRITE_CLASS_ENCODER(NotifyMessage);                        TYPE(librbd::trash_watcher::NotifyMessage)
src/librbd/watcher/Types.h:WRITE_CLASS_ENCODER(ClientId);
src/librbd/watcher/Types.h:WRITE_CLASS_ENCODER(NotifyResponse);
src/rbd_replay/ActionTypes.h:WRITE_CLASS_ENCODER(Dependency);                               TYPE(rbd_replay::action::Dependency)
src/rbd_replay/ActionTypes.h:WRITE_CLASS_ENCODER(ActionEntry);                              TYPE(rbd_replay::action::ActionEntry)
src/tools/rbd_mirror/image_map/Types.h:WRITE_CLASS_ENCODER(PolicyData);                     TYPE(rbd::mirror::image_map::PolicyData)
src/tools/rbd_mirror/instance_watcher/Types.h:WRITE_CLASS_ENCODER(NotifyMessage);
src/tools/rbd_mirror/instance_watcher/Types.h:WRITE_CLASS_ENCODER(NotifyAckPayload);
src/tools/rbd_mirror/leader_watcher/Types.h:WRITE_CLASS_ENCODER(NotifyMessage);
```

There are 57 WRITE_CLASS macros (58 - 1 duplicate) and 47 TYPE macros.  Out of those 47 types, only 14 are matching in the way they are specified: 4 are fully qualified, 10 aren't.  The majority (43 / 57 = 75%) are either not matching or missing completely.

Can you provide a list of types which are being tested correctly based on a dump from the corpus?

Reply:
@idryomov if we want to match the TYPE and WRITE_CLASS_ENCODER, that means we need to update all the corpus directories as well (if we updating the TYPE)

Reply:
> @idryomov if we want to match the TYPE and WRITE_CLASS_ENCODER, that means we need to update all the corpus directories as well (if we updating the TYPE)

It's been a while and the details are out of my head...  IIRC the reason I put together this table is that you stated that the namespace part of the name doesn't make it to the corpus:

> the corpuse repository contain types without namespaces. that causing ceph-dencoder to output the type is not recognize (wihtout full namespace), i didn't want to remove the duplicate exist type names.

But now, entries like `./objects/cls::rbd::MirrorPeer`, `./objects/librbd::journal::TagData`, etc in the column that you added today seem to contradict that -- the namespace part is clearly there.

So I'm not sure what you mean by "if we _want_ to match the TYPE and WRITE_CLASS_ENCODER" (emphasis mine).  Am I wrong in assuming that they _need_ to match for the tests to work correctly?

Reply:
> Am I wrong in assuming that they _need_ to match for the tests to work correctly?

If I'm not wrong, then a further observation is that the intersection of MATCH status and an entry being present in the corpus produces only two entries: `./objects/cls_rbd_parent` and `./objects/cls_rbd_snap`.
- Are these two types the only RBD-related types that are not only stored in the corpus and also checked?
- What is the behavior for entries such as `./objects/cls::rbd::MirrorPeer` which are stored in the corpus but their WRITE_CLASS_ENCODER and TYPE names don't match?  Are they just skipped?

Reply:
> Are they just skipped?

You may have answered this already in an earlier comment

> [...] since the corpus script first check if the type is supported by ceph-dencoder, its ignore all the types that return unsupported.

but I want to get a confirmation to avoid misunderstanding.

Is there some message logged when this happens?  If so, can you confirm that this message is produced for `cls::rbd::MirrorPeer` and all other types that are present in the corpus except for `cls_rbd_parent` and `cls_rbd_snap`?